File: android/src/main/java/com/audienzzrn/RNAudienzzModule.kt
package com.audienzz

/*
    Copyright 2025 Audienzz AG

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

*/

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.google.android.gms.ads.MobileAds
import org.audienzz.mobile.AudienzzPrebidMobile
import org.audienzz.mobile.AudienzzTargetingParams
import org.audienzz.mobile.api.data.AudienzzInitializationStatus
import org.audienzz.mobile.util.remote.RemoteConfigManager

private const val RN_SDK_VERSION = "0.4.4"

class RNAudienzzModule(reactContext: ReactApplicationContext) :
  ReactNativeModule(reactContext, SERVICE) {
  @ReactMethod
  fun initialize(companyID: String, promise: Promise) {
    AudienzzPrebidMobile.initializeSdk(applicationContext, companyID) { status ->
      when (status) {
        AudienzzInitializationStatus.SUCCEEDED -> {
          setupOmid()
          setupRnSdkIdentity()
          val result: WritableMap =
            Arguments.createMap().apply {
              putString("status", "SUCCEEDED")
              putString("description", "SDK initialized successfully!")
            }

          promise.resolve(result)
        }

        else -> {
          Log.e(TAG, "SDK initialization error: $status\n${status.description}")
        }
      }
    }
  }

  private fun setupOmid() {
    val v = MobileAds.getVersion()
    AudienzzTargetingParams.omidPartnerName = "Google"
    AudienzzTargetingParams.omidPartnerVersion = "${v.majorVersion}.${v.minorVersion}.${v.microVersion}"
  }

  private fun setupRnSdkIdentity() {
    AudienzzTargetingParams.setBridgeTargeting("au_rn_v", RN_SDK_VERSION)
  }

  /**
   * Supply a publisher-owned PPID (e.g. a hashed e-mail). Takes precedence over the SDK-generated
   * one; pass null to clear and fall back to it. A PPID is always sent — there is no opt-out.
   */
  @ReactMethod
  fun setPublisherPpid(ppid: String?) {
    AudienzzPrebidMobile.ppidManager?.setPublisherPpid(ppid)
  }

  @ReactMethod
  fun getPpid(promise: Promise) {
    promise.resolve(AudienzzPrebidMobile.ppidManager?.getPpid())
  }

  @ReactMethod
  fun setSchainObject(schain: String) {
    AudienzzPrebidMobile.setSchainObject(schain)
  }

  /** One greppable AUDZ line per slot decision; see AudienzzDiagnostics. */
  @ReactMethod
  fun setDiagnosticsEnabled(enabled: Boolean) {
    AudienzzPrebidMobile.diagnosticsEnabled = enabled
  }

  /**
   * Force smart-refresh v2 on/off, overriding the backend `smartRefreshV2` config for the session.
   * v2 uses the directional viewport gate; v1 uses the legacy >=20%-visible gate.
   */
  @ReactMethod
  fun setSmartRefreshV2Enabled(enabled: Boolean) {
    AudienzzPrebidMobile.smartRefreshV2Override = enabled
  }

  /** When true, a banner blanks its slot during a screen-resume reload. */
  @ReactMethod
  fun setBlankOnScreenReload(enabled: Boolean) {
    AudienzzPrebidMobile.blankOnScreenReload = enabled
  }

  /** Global GMA ad audio volume for all ad types. Clamped to [0,1]; 0 = muted. */
  @ReactMethod
  fun setAppVolume(volume: Float) {
    AudienzzPrebidMobile.setAppVolume(volume.coerceIn(0f, 1f))
  }

  /**
   * Report an ad-bearing screen, dialog, or popup by name (your JS navigation route). Fires a
   * `pageImpression` and starts a fresh page-impression id that ties all ad events on this screen
   * visit together. Call on each navigation to an ad-bearing screen.
   */
  @ReactMethod
  fun pageImpression(name: String) {
    AudienzzPrebidMobile.pageImpression(name)
  }

  /**
   * Report a page whose identity and analytics name differ. Every React Native ad lives in the one
   * host Activity, so host identity can never separate two routes — the id is the only thing that
   * can, and a screen name repeats (two articles are both "article").
   */
  @ReactMethod
  fun pageImpressionWithId(pageId: String, name: String) {
    AudienzzPrebidMobile.pageImpression(pageId, name)
  }

  /**
   * Our own registration, so teardown removes ONLY ours and never someone else's.
   *
   * Forwards every native page impression to JS -- including the automatic one fired on returning
   * to the foreground, which never passes through the JS API. Native owns foreground reporting; JS
   * just page-scopes the ad types the native coordinator doesn't track (rendering banners).
   */
  private var pageImpressionObserver: ((String) -> Unit)? = { name ->
    reactContext
      .getJSModule(com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(PAGE_IMPRESSION_EVENT, name)
  }

  init {
    AudienzzPrebidMobile.pageImpressionObserver = pageImpressionObserver
  }

  override fun invalidate() {
    // The observer is a process-global singleton capturing the ReactApplicationContext. Without
    // this, destroying the React instance leaves that context retained and later page impressions
    // still emitting into a dead bridge.
    if (AudienzzPrebidMobile.pageImpressionObserver === pageImpressionObserver) {
      AudienzzPrebidMobile.pageImpressionObserver = null
    }
    pageImpressionObserver = null
    super.invalidate()
  }

  @ReactMethod
  fun configureRemote(remoteUrl: String, publisherId: String, promise: Promise) {
    try {
      RemoteConfigManager.initialize(
        publisherId = publisherId,
        remoteUrl = remoteUrl
      )
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("CONFIGURE_REMOTE_FAILED", e.message, e)
    }
  }

  @ReactMethod
  fun fetchPublisherConfig(publisherId: String, promise: Promise) {
    AudienzzPrebidMobile.initializeRemoteSdk(
      applicationContext,
      publisherId
    ) { status ->
      if (status == AudienzzInitializationStatus.SUCCEEDED) {
        setupOmid()
        setupRnSdkIdentity()
        promise.resolve(null)
      } else {
        val description = status.description ?: "Remote configuration fetch failed"
        promise.reject("FETCH_FAILED", description)
      }
    }
  }

  @ReactMethod
  fun getStickyConfig(adConfigId: String, promise: Promise) {
    AudienzzPrebidMobile.getAdUnitConfig(adConfigId) { config ->
      val result = Arguments.createMap().apply {
        putInt("maxHeight", config?.config?.stickyMaxHeight ?: 600)
        putInt("stickyTopOffset", config?.config?.stickyTopOffset ?: 0)
      }
      promise.resolve(result)
    }
  }

  companion object {
    private const val SERVICE = "RNAudienzzModule"
    private const val TAG = "AudienzzSDKInitializer"

    /** Device event carrying the page name of every native page impression. */
    const val PAGE_IMPRESSION_EVENT = "AudienzzPageImpression"
  }
}

File: src/types/RNAudienzzModule.ts
export type AudienzzInitStatus = {
  status: string;
  description: string;
};

export interface RNAudienzzModule {
  initialize(companyId: string): Promise<AudienzzInitStatus>;

  setPublisherPpid(ppid: string | null): void;
  getPpid(): Promise<string | null>;

  setSchainObject(schain: string): Promise<void>;

  /**
   * Set the global GMA ad audio volume for all ad types. `volume` is clamped to [0.0, 1.0]
   * (0.0 = muted, 1.0 = full device volume). The SDK defaults to muted on init.
   */
  setAppVolume(volume: number): void;

  /**
   * Force smart-refresh v2 on/off, overriding the backend `smartRefreshV2` config for the session.
   * v2 uses the directional viewport gate; v1 uses the legacy ≥20%-visible gate. Omit to defer to backend.
   */
  /** Mirror the JS diagnostics switch into the native SDK. See `Audienzz.setDiagnosticsEnabled`. */
  setDiagnosticsEnabled(enabled: boolean): void;
  setSmartRefreshV2Enabled(enabled: boolean): void;

  /**
   * When true, a banner blanks its slot during a screen-resume reload (native `blankOnScreenReload`).
   * Default false — the old creative stays until the new one arrives.
   */
  setBlankOnScreenReload(enabled: boolean): void;

  /**
   * Report an ad-bearing screen, dialog, or popup by name (your navigation route id/name). Fires a
   * `pageImpression` and starts a fresh page-impression id that ties all ad events on this screen
   * visit together. Call on each navigation to an ad-bearing screen.
   */
  pageImpression(name: string): void;

  /**
   * Report a page whose identity and analytics name differ. `pageId` identifies the route instance
   * and is what banners are matched against; `name` is what analytics records.
   */
  pageImpressionWithId(pageId: string, name: string): void;

  configureRemote(remoteUrl: string, publisherId: string): Promise<void>;
  fetchPublisherConfig(publisherId: string): Promise<void>;

  /**
   * Initialize SDK with remote configuration
   * This method combines configureRemote() and fetchPublisherConfig().
   * @param remoteUrl - The remote configuration API URL. Android 0.3.0 supports only
   * https://api.adnz.co/api/ws-sdk-config/public/v1/ and rejects other endpoints.
   * @param publisherId - The publisher ID
   */
  initializeRemote(remoteUrl: string, publisherId: string): Promise<AudienzzInitStatus>;

  /**
   * Returns the backend-configured sticky wrapper dimensions for a given ad config ID.
   * Values are resolved as: backend value → SDK default (maxHeight: 600, stickyTopOffset: 0).
   * Used internally by `AudienzzStickyAdWrapper` when `adConfigId` is provided.
   */
  getStickyConfig(adConfigId: string): Promise<{ maxHeight: number; stickyTopOffset: number }>;
}
